train.py

```diff
@@ -107,10 +107,35 @@ def main(args):
         # save checkpoint
         if epoch_itr.epoch % args.save_interval == 0:
             save_checkpoint(args, trainer, epoch_itr, valid_losses[0])
+
+        epoch_itr = reload_train(args, epoch_itr, max_positions, task)
     train_meter.stop()
     print('| done training in {:.1f} seconds'.format(train_meter.sum))
 
 
+def reload_train(args, epoch_itr, max_positions, task):
+    # nothing needs to be done when the dataset is not sharded.
+    if len(args.data.split(":")) == 1:
+        return epoch_itr
+    print("| Reloading shard of train data at epoch: ", epoch_itr.epoch)
+    load_dataset_splits(args, task, [args.train_subset], epoch=epoch_itr.epoch)
+    epoch_state_dict = epoch_itr.state_dict()
+    epoch_itr = task.get_batch_iterator(
+        dataset=task.dataset(args.train_subset),
+        max_tokens=args.max_tokens,
+        max_sentences=args.max_sentences,
+        max_positions=max_positions,
+        ignore_invalid_inputs=True,
+        required_batch_size_multiple=args.required_batch_size_multiple,
+        seed=args.seed,
+        num_shards=args.distributed_world_size,
+        shard_id=args.distributed_rank,
+        num_workers=args.num_workers,
+    )
+    epoch_itr.load_state_dict(epoch_state_dict)
```

Comment:
I'm still not sure about hacking the state dict like this. All we need is to set the initial epoch value right? Maybe we can pass an epoch= kwarg to get_batch_iterator so that EpochBatchIterator won't default epoch to 0: https://github.com/pytorch/fairseq/blob/master/fairseq/data/iterators.py#L97